src/dbnode/storage/repair.go

```diff
@@ -198,33 +223,66 @@ func (r shardRepairer) Repair(
 			}
 
 			for _, replicaMetadata := range replicaMetadataBlocks.Metadata() {
-				if replicaMetadata.Host.ID() == session.Origin().ID() {
+				metadataHostID := replicaMetadata.Host.ID()
+				if metadataHostID == originID {
 					// Don't request blocks for self metadata.
 					continue
 				}
-				metadatasToFetchBlocksFor = append(metadatasToFetchBlocksFor, replicaMetadata)
+
+				if len(sessions) == 1 {
+					// Optimized path for single session case.
+					metadatasToFetchBlocksForPerSession[0] = append(metadatasToFetchBlocksForPerSession[0], replicaMetadata)
+					continue
+				}
+
+				// If there is more than one session then we need to match up all of the metadata to the
+				// session it belongs to so that we can fetch the corresponding blocks of data.
+				for i, sesTopo := range sessions {
+					_, ok := sesTopo.topo.LookupHostShardSet(metadataHostID)
+					if !ok {
+						// The host this metadata came from is not part of the cluster this session is connected to.
+						continue
+					}
+					metadatasToFetchBlocksForPerSession[i] = append(metadatasToFetchBlocksForPerSession[i], replicaMetadata)
+					break
+				}
+
+				// Could happen during topology changes (I.E node is kicked out of the cluster in-between
+				// fetching its metadata and this step).
+				r.logger.Debug(
```

Comment:
Isn't this guaranteed to run if `len(sessions) > 1`?

Reply:
you're right I need a variable to track, thanks